src/Orleans/Utils/IKeyedServiceCollection.cs

```diff
@@ -10,7 +12,40 @@ namespace Orleans.Runtime
     public interface IKeyedServiceCollection<in TKey, out TService>
         where TService : class
     {
-        TService GetService(TKey key);
+        TService GetService(IServiceProvider services, TKey key);
+    }
+
+    public interface IKeyedService<TKey, out TService> : IEquatable<TKey>
+    {
+        TService GetService(IServiceProvider services);
+    }
+
+    public class KeyedService<TKey, TService, TInstance> : IKeyedService<TKey, TService>
+            where TInstance : TService
+    {
+        private readonly TKey key;
+
+        public KeyedService(TKey key)
+        {
+            this.key = key;
+        }
+
+        public TService GetService(IServiceProvider services) => services.GetService<TInstance>();
```

Comment:
BTW, just talked in person, but mentioning it here for others: I'm not suggesting switching to Scoped... Transient is probably better, otherwise we would hold a reference to the factory during the 2 hours that the grain is activated (or whatever the lifetime of the grain is) for no reason. It's better to use the factories when activating and just garbage collect them afterwards.